Enable integration of Exiv2 as CMake sub-project

Fimagena opened the pull request:
When integrating Exiv2 into another CMake project with add_subdirectory() the CMake-run fails on FindIconv, because the CMAKE_MODULE_PATH points to the top-level project, not to the Exiv2-project level.

Fimagena committed:
Enable integration of Exiv2 as CMake sub-project

When integrating Exiv2 into another CMake project with add_subdirectory() the CMake-run fails on FindIconv, because the CMAKE_MODULE_PATH points to the top-level project, not to the Exiv2-project level.

clanmills commented:
Thanks @Fimagena for this PR.  @piponazo is Exiv2's CMake Expert and should review your change.

Fimagena committed:
Use relative references in cmake-files and make libxmp private

Relative references ensure that exiv2 can be integrated more easily in other CMake-projects. If there is no specific reason for libxmp to export it symbols, it should be private.

Fimagena committed:
Make XMP CMake-include reference relative

clanmills commented:
If you are linking both the Adobe XMPsdk and the xmpsdk source included in Exiv2, you will have linking errors.  The Adobe code uses externs which cause linking errors.  We have a work-in-progress project to build and link "pure" Adobe XMPsdk and Exiv2 can be built to use one of three editions: 2016|2014|2013, or the existing xmpsdk source, or omitted entirely.  At the moment, building with an external Adobe XMPsdk is only implemented using ./configure (autotools), although I believe the changes required to support this from CMake are small.

piponazo commented on src/CMakeLists.txt:
```diff
@@ -204,18 +204,18 @@ endif()
 if ( EXIV2_ENABLE_XMP )
     #target_include_directories(exiv2lib PRIVATE ${CMAKE_SOURCE_DIR}/xmpsdk/include)
-    target_link_libraries(exiv2lib PUBLIC xmp)
+    target_link_libraries(exiv2lib PRIVATE xmp)
```
I think this is the culprit of the linking error detected by AppVeyor:

```
[154/168] Linking CXX executable bin\geotag.exe
FAILED: bin/geotag.exe 
cmd.exe /C "cd . && "C:\Program Files (x86)\CMake\bin\cmake.exe" -E vs_link_exe --intdir=samples\CMakeFiles\geotag.dir --manifests  -- C:\PROGRA~2\MI0E91~1.0\VC\bin\amd64\link.exe /nologo samples\CMakeFiles\geotag.dir\geotag.cpp.obj  /out:bin\geotag.exe /implib:lib\geotag.lib /pdb:bin\geotag.pdb /version:0.0  /machine:x64 /INCREMENTAL:NO  /subsystem:console -LIBPATH:C:\projects\exiv2-wutfp\build\install\lib lib\exiv2.lib C:\Users\appveyor\conanCache\zlib\1.2.11\lasote\stable\package\63da998e3642b50bee33f4449826b2d623661505\lib\zlibstatic.lib C:\.conan\fvzhjf\1\lib\libcurl_imp.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK: command "C:\PROGRA~2\MI0E91~1.0\VC\bin\amd64\link.exe /nologo samples\CMakeFiles\geotag.dir\geotag.cpp.obj /out:bin\geotag.exe /implib:lib\geotag.lib /pdb:bin\geotag.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console -LIBPATH:C:\projects\exiv2-wutfp\build\install\lib lib\exiv2.lib C:\Users\appveyor\conanCache\zlib\1.2.11\lasote\stable\package\63da998e3642b50bee33f4449826b2d623661505\lib\zlibstatic.lib C:\.conan\fvzhjf\1\lib\libcurl_imp.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:bin\geotag.exe.manifest" failed (exit code 1120) with the following output:
geotag.cpp.obj : error LNK2019: unresolved external symbol XML_ParserCreate referenced in function "bool __cdecl readXML(char const *,class Options &)" (?readXML@@YA_NPEBDAEAVOptions@@@Z)
geotag.cpp.obj : error LNK2019: unresolved external symbol XML_SetElementHandler referenced in function "bool __cdecl readXML(char const *,class Options &)" (?readXML@@YA_NPEBDAEAVOptions@@@Z)
geotag.cpp.obj : error LNK2019: unresolved external symbol XML_SetCharacterDataHandler referenced in function "bool __cdecl readXML(char const *,class Options &)" (?readXML@@YA_NPEBDAEAVOptions@@@Z)
geotag.cpp.obj : error LNK2019: unresolved external symbol XML_SetUserData referenced in function "bool __cdecl readXML(char const *,class Options &)" (?readXML@@YA_NPEBDAEAVOptions@@@Z)
geotag.cpp.obj : error LNK2019: unresolved external symbol XML_Parse referenced in function "bool __cdecl readXML(char const *,class Options &)" (?readXML@@YA_NPEBDAEAVOptions@@@Z)
geotag.cpp.obj : error LNK2019: unresolved external symbol XML_ParserFree referenced in function "bool __cdecl readXML(char const *,class Options &)" (?readXML@@YA_NPEBDAEAVOptions@@@Z)
bin\geotag.exe : fatal error LNK1120: 6 unresolved externals
```
Some of the sample applicaitons are using XMP symbols directly and we were exposing them through the *exiv2lib* module. Maybe we can keep this private and link against xmp directly in the sample applications depending on it. Do you have a windows workstation where to try those changes? Otherwise I can try it by myself.

piponazo reviewed (changes requested):
Thanks @Fimagena for this contribution. In principle I am not against this change, if it can add the ability to integrate exiv2 into a existing project as a CMake-subproject without breaking the normal workflow.

However, Appveyor is reporting a linking error right now during the compilation of a sample application on windows. Please, let me know if you can try to solve the issue on windows, otherwise I can give you a hand.

D4N commented:
@Fimagena Do you have the resources to address the build issues on Windows or do you want help with that?

D4N commented:
Now that #462 has been merged and this PR has been stale for nearly half a year, I propose that we close it.

piponazo commented:
Right, this can be closed now.
